pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java

```diff
@@ -757,57 +759,181 @@ private void createDerivedColumnV1Indices(String column, FunctionEvaluator funct
+        targetMaxChunkSizeBytes = forwardIndexConfig.getTargetMaxChunkSizeBytes();
+        targetDocsPerChunk = forwardIndexConfig.getTargetDocsPerChunk();
       }
-    } finally {
-      for (ValueReader valueReader : valueReaders) {
-        valueReader.close();
+    }
+    if (chunkCompressionType == null) {
+      chunkCompressionType = ForwardIndexType.getDefaultCompressionType(fieldSpec.getFieldType());
+    }
+
+    if (isSingleValue) {
+      try (ForwardIndexCreator forwardIndexCreator = fieldSpec.getDataType().getStoredType().isFixedWidth()
+          ? new SingleValueFixedByteRawIndexCreator(_indexDir, chunkCompressionType, column, numDocs,
+              fieldSpec.getDataType().getStoredType(), rawIndexWriterVersion, targetDocsPerChunk)
+          : new SingleValueVarByteRawIndexCreator(_indexDir, chunkCompressionType, column, numDocs,
+              fieldSpec.getDataType().getStoredType(), indexCreationInfo.getMaxRowLengthInBytes(),
+              deriveNumDocsPerChunk, rawIndexWriterVersion, targetMaxChunkSizeBytes, targetDocsPerChunk)) {
+        for (Object outputValue : outputValues) {
+          switch (fieldSpec.getDataType().getStoredType()) {
+            // Casts are safe here because we've already done the conversion in createDerivedColumnV1Indices
+            case INT:
+              forwardIndexCreator.putInt((int) outputValue);
+              break;
+            case LONG:
+              forwardIndexCreator.putLong((long) outputValue);
+              break;
+            case FLOAT:
+              forwardIndexCreator.putFloat((float) outputValue);
+              break;
+            case DOUBLE:
+              forwardIndexCreator.putDouble((double) outputValue);
+              break;
+            case BIG_DECIMAL:
+              forwardIndexCreator.putBigDecimal((BigDecimal) outputValue);
+              break;
+            case STRING:
+              forwardIndexCreator.putString((String) outputValue);
+              break;
+            case BYTES:
+              forwardIndexCreator.putBytes((byte[]) outputValue);
+              break;
+            default:
+              throw new IllegalStateException();
+          }
+        }
+      }
+    } else {
+      try (ForwardIndexCreator forwardIndexCreator = fieldSpec.getDataType().getStoredType().isFixedWidth()
+          ? new MultiValueFixedByteRawIndexCreator(_indexDir, chunkCompressionType, column, numDocs,
+              fieldSpec.getDataType().getStoredType(), indexCreationInfo.getMaxNumberOfMultiValueElements(),
+              deriveNumDocsPerChunk, rawIndexWriterVersion, targetMaxChunkSizeBytes, targetDocsPerChunk)
+          : new MultiValueVarByteRawIndexCreator(_indexDir, chunkCompressionType, column, numDocs,
+              fieldSpec.getDataType().getStoredType(), rawIndexWriterVersion,
+              indexCreationInfo.getMaxRowLengthInBytes(), indexCreationInfo.getMaxNumberOfMultiValueElements(),
+              targetMaxChunkSizeBytes, targetDocsPerChunk)) {
```

Comment:
This is starting to look fairly messy, I'm wondering if we should refactor to use [ForwardIndexCreatorFactory](https://github.com/apache/pinot/blob/e07b576797e2dc834a60d8ee0da5d26a0b597e23/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/forward/ForwardIndexCreatorFactory.java#L47) instead?